uPortal-soffit/uPortal-soffit-core/src/main/java/org/apereo/portal/soffit/service/JwtEncryptor.java

```diff
@@ -0,0 +1,97 @@
+import java.util.regex.Pattern;
+import javax.annotation.PostConstruct;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.jasypt.util.text.BasicTextEncryptor;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+/**
+ * Wrapper for code to encrypt and decrypt JWT tokens. The spec is evolving, so this implementation
+ * is expected to change.
+ *
+ * @since 5.6.1
+ */
+@Slf4j
+@Component
+public class JwtEncryptor {
+    public static final String ENCRYPT_JWT_YN = "org.apereo.portal.soffit.jwt.encrypt";
+    public static final String DEFAULT_ENCRYPT_JWT_YN = "true";
+
+    public static final String ENCRYPTION_PASSWORD_PROPERTY =
+            "org.apereo.portal.soffit.jwt.encryptionPassword";
+    public static final String DEFAULT_ENCRYPTION_PASSWORD = "CHANGEME";
+
+    @Value("${" + ENCRYPTION_PASSWORD_PROPERTY + ":" + DEFAULT_ENCRYPTION_PASSWORD + "}")
+    private String encryptionPassword;
+
+    // BASE64url encoding, per RFC7519
+    public static final String BASE64_REGEX = "[\\p{Alnum}-_=]+";
+    public static final Pattern JWT_PATTERN =
+            Pattern.compile("^" + BASE64_REGEX + "\\." + BASE64_REGEX + "\\." + BASE64_REGEX + "$");
+
+    @Value("${" + ENCRYPT_JWT_YN + ":" + DEFAULT_ENCRYPT_JWT_YN + "}")
+    private boolean encryptJwtYN;
+
+    /*
+     * NOTE:  There is also a StrongTextEncryptor, but it requires each deployment
+     * to download and install the "Java Cryptography Extension (JCE) Unlimited
+     * Strength Jurisdiction Policy Files," which sounds like a tremendous PITA.
+     * The BasicTextEncryptor supports "normal-strength encryption of texts,"
+     * which should be satisfactory for our needs.
+     */
+    final BasicTextEncryptor textEncryptor = new BasicTextEncryptor();
+
+    @PostConstruct
+    public void init() {
+        if (encryptJwtYN) {
+            // log.info("Encrypting Soffit JWTs");
+            // Encryption Password
+            if (StringUtils.isBlank(encryptionPassword)) {
+                log.error(
+                        "The value of required property {} is blank", ENCRYPTION_PASSWORD_PROPERTY);
+                throw new IllegalStateException("Missing property " + ENCRYPTION_PASSWORD_PROPERTY);
+            } else if (DEFAULT_ENCRYPTION_PASSWORD.equals(encryptionPassword)) {
+                log.warn(
+                        "Property {} is using the deafult value;  please change it",
+                        ENCRYPTION_PASSWORD_PROPERTY);
+            }
+            textEncryptor.setPassword(encryptionPassword);
+        }
+    }
+
+    public String encryptIfConfigured(String jwt) {
+        if (encryptJwtYN) {
+            return textEncryptor.encrypt(jwt);
+        } else {
+            return jwt;
+        }
+    }
+
+    public String decryptIfTokenEncrypted(String jwt) {
+        if (encryptJwtYN) {
+            Matcher m = JWT_PATTERN.matcher(jwt);
```

Comment:
Does the JWT library used in uPortal include a util function that checks if a string is a JWT?

Reply:
I did not find a format validator; however, I did look at their code to see how the validate the format. This whole issue was due to this JJWT code throwing an exception because there were not 2 `.`s in the string.

Reply:
Would there be any negatives to going with something like
```
try {
  decodeJWT
} catch {
  decryptJWT
  decodeJWT
}
```

to avoid maintaining a separate regex?

Reply:
Yes, 1) depending on an exception when a known condition is expected is a bad practice, 2) the regex follows a well-defined spec. Also this special handling will go away in the next major release. It's here to deal with Soffit and OIDC using the same header for their potentially different tokens.

Reply:
> It's here to deal with Soffit and OIDC using the same header for their potentially different tokens.

If OIDC and Soffit are using different header handling mechanisms, should this class get split, one for soffit, one for OIDC?

> special handling will go away in the next major release

My caution is without a comment or having a separate class, it would be easy to forget and miss that this should be removed in the next release.

Reply:
> depending on an exception when a known condition is expected is a bad practice

It probably emits a known exception.
Would

```java
try {
  decodeJWT(); // might be encrypted, if so it will throw
} catch (JwtDecodeError e) { // catch only decode exception
  decryptJWT();
  decodeJWT();
} catch (Error e) {
  throw e; // rethrow other errors
}
```